File: app/src/main/java/com/datarockets/mnchkn/store/MunchkinDatabaseHelper.java
package com.datarockets.mnchkn.store;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.datarockets.mnchkn.models.GameStep;
import com.datarockets.mnchkn.models.Player;
import com.datarockets.mnchkn.utils.LogUtil;

import java.util.ArrayList;

public class MunchkinDatabaseHelper extends SQLiteOpenHelper {

    private static final String TAG = LogUtil.makeLogTag(MunchkinDatabaseHelper.class);

    private static final String DATABASE_NAME = "players_db";
    private static final int DATABASE_VERSION = 2;

    private static final String TABLE_PLAYERS = "players";

    private static final String KEY_PLAYER_ID = "id";
    private static final String KEY_PLAYER_NAME = "name";
    private static final String KEY_PLAYER_LEVEL = "level";
    private static final String KEY_PLAYER_STRENGTH = "strength";
    private static final String KEY_PLAYER_COLOR = "color";
    private static final String KEY_PLAYER_TOTAL = "total";

    private static final String TABLE_GAME = "game";

    private static final String KEY_GAME_PLAYER_ID = "player_id";
    private static final String KEY_GAME_PLAYER_LEVEL = "player_level";
    private static final String KEY_GAME_PLAYER_STRENGTH = "player_strength";

    private static final String ORDER_BY = " ORDER BY ";

    private static final int ORDER_BY_ID = 0;
    private static final int ORDER_BY_LEVEL = 1;
    private static final int ORDER_BY_STRENGTH = 2;
    private static final int ORDER_BY_TOTAL = 3;

    private static MunchkinDatabaseHelper instance;

    public static synchronized MunchkinDatabaseHelper getInstance(Context context) {
        if (instance == null) {
            instance = new MunchkinDatabaseHelper(context.getApplicationContext());
        }
        return instance;
    }

    private MunchkinDatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String createPlayersTableQuery = "CREATE TABLE " + TABLE_PLAYERS +
                "(" +
                KEY_PLAYER_ID + " INTEGER PRIMARY KEY," +
                KEY_PLAYER_NAME + " TEXT," +
                KEY_PLAYER_LEVEL + " INTEGER," +
                KEY_PLAYER_STRENGTH + " INTEGER," +
                KEY_PLAYER_COLOR + " TEXT" +
                ")";
        String createGameTableQuery = "CREATE TABLE " + TABLE_GAME +
                "(" +
                KEY_GAME_PLAYER_ID + " INTEGER REFERENCES " + TABLE_PLAYERS + "," +
                KEY_GAME_PLAYER_LEVEL + " INTEGER," +
                KEY_GAME_PLAYER_STRENGTH + " INTEGER" +
                ")";
        db.execSQL(createPlayersTableQuery);
        db.execSQL(createGameTableQuery);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        switch (oldVersion) {
            case 1:
                db.execSQL("ALTER TABLE "
                        + TABLE_PLAYERS + " ADD COLUMN "
                        + KEY_PLAYER_COLOR + " TEXT");
                if (!isTableEmpty(db, TABLE_PLAYERS)) {
                    addColorsToUpdatedPlayers(db);
                }
            default:
                break;
        }
    }

    public long addPlayer(Player player) {
        long playerId = -1;
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            ContentValues values = new ContentValues();
            values.put(KEY_PLAYER_NAME, player.name);
            values.put(KEY_PLAYER_LEVEL, player.levelScore);
            values.put(KEY_PLAYER_STRENGTH, player.strengthScore);
            values.put(KEY_PLAYER_COLOR, player.color);
            playerId = db.insertOrThrow(TABLE_PLAYERS, null, values);
            db.setTransactionSuccessful();
            Log.i(TAG, "Player id is " + playerId);
        } catch (Exception e) {
            Log.e(TAG, "Error while trying to add new player");
        } finally {
            db.endTransaction();
        }
        return playerId;
    }

    public ArrayList<Player> getPlayers() {
        return getPlayers(ORDER_BY_ID);
    }

    public ArrayList<Player> getPlayers(int orderValue) {
        ArrayList<Player> playersList = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        String orderByQuery = null;
        switch (orderValue) {
            case ORDER_BY_ID:
                orderByQuery = ORDER_BY + KEY_PLAYER_ID;
                break;
            case ORDER_BY_LEVEL:
                orderByQuery = ORDER_BY + KEY_PLAYER_LEVEL;
                break;
            case ORDER_BY_STRENGTH:
                orderByQuery = ORDER_BY + KEY_PLAYER_STRENGTH;
                break;
            case ORDER_BY_TOTAL:
                orderByQuery = ORDER_BY + KEY_PLAYER_TOTAL;
                break;
        }
        String playersTotalQuery = "SELECT "
                + KEY_PLAYER_ID + ", "
                + KEY_PLAYER_NAME + ", "
                + KEY_PLAYER_LEVEL + ", "
                + KEY_PLAYER_STRENGTH + ", "
                + KEY_PLAYER_COLOR + ", ("
                + KEY_PLAYER_LEVEL + " + " + KEY_PLAYER_STRENGTH + ") AS " + KEY_PLAYER_TOTAL
                + " FROM " + TABLE_PLAYERS + orderByQuery + " DESC";
        Log.v(TAG, playersTotalQuery);
        Cursor cursor = db.rawQuery(playersTotalQuery, null);
        try {
            if (cursor.moveToFirst()) {
                do {
                    Player player = new Player();
                    player.setId(cursor.getLong(cursor.
                            getColumnIndex(KEY_PLAYER_ID)));
                    player.setName(cursor.getString(cursor.
                            getColumnIndex(KEY_PLAYER_NAME)));
                    player.setColor(cursor.getString(cursor.
                            getColumnIndex(KEY_PLAYER_COLOR)));
                    player.setLevelScore(cursor.getInt(cursor.
                            getColumnIndex(KEY_PLAYER_LEVEL)));
                    player.setStrengthScore(cursor.getInt(cursor.
                            getColumnIndex(KEY_PLAYER_STRENGTH)));
                    player.setTotalScore(cursor.getInt(cursor.
                            getColumnIndex(KEY_PLAYER_TOTAL)));
                    playersList.add(player);
                } while (cursor.moveToNext());
            }
        } catch (Exception e) {
            Log.e(TAG, "Error while getting players");
        } finally {
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return playersList;
    }

    public Player updatePlayer(Player player) {
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(KEY_PLAYER_LEVEL, player.levelScore);
        values.put(KEY_PLAYER_STRENGTH, player.strengthScore);
        db.update(TABLE_PLAYERS, values, KEY_PLAYER_ID + " = ?",
                new String[] {String.valueOf(player.id)});
        return player;
    }

    public void deletePlayer(long id) {
        Log.v(TAG, "Id of the player is " + id);
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            db.delete(TABLE_PLAYERS, KEY_PLAYER_ID + " = ?", new String[] {Long.toString(id)});
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e(TAG, "Error while deleting player");
        } finally {
            db.endTransaction();
        }
    }

    private boolean isTableEmpty(SQLiteDatabase db, String tableName) {
        String countQuery = "SELECT COUNT(*) FROM " + tableName;
        Cursor cursor = db.rawQuery(countQuery, null);
        cursor.moveToFirst();
        int count = cursor.getInt(0);
        if (count > 0) {
            return false;
        } else {
            return true;
        }
    }

    private void addColorsToUpdatedPlayers(SQLiteDatabase db) {
        ContentValues values = new ContentValues();
        values.put(KEY_PLAYER_COLOR, "#3B1606");
        db.update(TABLE_PLAYERS, values, null, null);
    }

    public void clearPlayersStats() {
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(KEY_PLAYER_LEVEL, 1);
        values.put(KEY_PLAYER_STRENGTH, 1);
        db.update(TABLE_PLAYERS, values, null, null);
    }

    public void insertStep(GameStep gameStep) {
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            ContentValues values = new ContentValues();
            values.put(KEY_GAME_PLAYER_ID, gameStep.getPlayerId());
            values.put(KEY_GAME_PLAYER_LEVEL, gameStep.getPlayerLevel());
            values.put(KEY_GAME_PLAYER_STRENGTH, gameStep.getPlayerStrength());
            db.insertOrThrow(TABLE_GAME, null, values);
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e(TAG, "Error while trying to insert step to database");
        } finally {
            db.endTransaction();
        }
    }

    public void clearSteps() {
        Log.v(TAG, "Clearing game steps");
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            db.delete(TABLE_GAME, null, null);
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e(TAG, "Error while trying to delete game steps");
        } finally {
            db.endTransaction();
        }
    }

    public ArrayList<GameStep> getGameSteps() {
        Log.v(TAG, "GET GAME STEPS");
        ArrayList<GameStep> gameSteps = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        String lineChartQuery = "SELECT "
                + KEY_GAME_PLAYER_ID + ", "
                + KEY_GAME_PLAYER_LEVEL + ", "
                + KEY_GAME_PLAYER_STRENGTH + ", "
                + KEY_PLAYER_NAME + ", "
                + KEY_PLAYER_COLOR + " FROM " + TABLE_GAME
                + " INNER JOIN " + TABLE_PLAYERS + " ON players.id = " + KEY_GAME_PLAYER_ID;
        Log.v("TAG", lineChartQuery);
        Cursor cursor = db.rawQuery(lineChartQuery, null);
        try {
            if (cursor.moveToFirst()) {
                do {
                    GameStep gameStep = new GameStep();
                    gameStep.setPlayerId(cursor.getLong(cursor.
                            getColumnIndex(KEY_GAME_PLAYER_ID)));
                    gameStep.setPlayerLevel(cursor.getInt(cursor.
                            getColumnIndex(KEY_GAME_PLAYER_LEVEL)));
                    gameStep.setPlayerStrength(cursor.getInt(cursor.
                            getColumnIndex(KEY_GAME_PLAYER_STRENGTH)));
                    gameSteps.add(gameStep);
                } while (cursor.moveToNext());
            }
        } catch (Exception e) {
            Log.e(TAG, "Error while getting game steps");
        } finally {
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return gameSteps;
    }

}

File: app/src/main/kotlin/com/datarockets/mnchkn/store/MunchkinDatabaseHelper.kt
package com.datarockets.mnchkn.store

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import com.datarockets.mnchkn.models.GameStep
import com.datarockets.mnchkn.models.Player
import com.datarockets.mnchkn.utils.LogUtil
import java.util.*
import javax.inject.Singleton

@Singleton
class MunchkinDatabaseHelper(context: Context) : SQLiteOpenHelper(context, MunchkinDatabaseHelper.DATABASE_NAME, null, MunchkinDatabaseHelper.DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase) {
        val createPlayersTableQuery = "CREATE TABLE " + TABLE_PLAYERS +
                "(" +
                KEY_PLAYER_ID + " INTEGER PRIMARY KEY," +
                KEY_PLAYER_NAME + " TEXT," +
                KEY_PLAYER_LEVEL + " INTEGER," +
                KEY_PLAYER_STRENGTH + " INTEGER," +
                KEY_PLAYER_COLOR + " TEXT" +
                ")"
        val createGameTableQuery = "CREATE TABLE " + TABLE_GAME +
                "(" +
                KEY_GAME_PLAYER_ID + " INTEGER REFERENCES " + TABLE_PLAYERS + "," +
                KEY_GAME_PLAYER_LEVEL + " INTEGER," +
                KEY_GAME_PLAYER_STRENGTH + " INTEGER" +
                ")"
        db.execSQL(createPlayersTableQuery)
        db.execSQL(createGameTableQuery)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        when (oldVersion) {
            1 -> {
                db.execSQL("ALTER TABLE "
                        + TABLE_PLAYERS + " ADD COLUMN "
                        + KEY_PLAYER_COLOR + " TEXT")
                if (!isTableEmpty(db, TABLE_PLAYERS)) {
                    addColorsToUpdatedPlayers(db)
                }
            }
            else -> {
            }
        }
    }

    fun addPlayer(player: Player): Long {
        var playerId: Long = -1
        val db = writableDatabase
        db.beginTransaction()
        try {
            val values = ContentValues()
            values.put(KEY_PLAYER_NAME, player.name)
            values.put(KEY_PLAYER_LEVEL, player.levelScore)
            values.put(KEY_PLAYER_STRENGTH, player.strengthScore)
            values.put(KEY_PLAYER_COLOR, player.color)
            playerId = db.insertOrThrow(TABLE_PLAYERS, null, values)
            db.setTransactionSuccessful()
            Log.i(TAG, "Player id is " + playerId)
        } catch (e: Exception) {
            Log.e(TAG, "Error while trying to add new player")
        } finally {
            db.endTransaction()
        }
        return playerId
    }

    val players: ArrayList<Player>
        get() = getPlayers(ORDER_BY_ID)

    fun getPlayers(orderValue: Int): ArrayList<Player> {
        val playersList = ArrayList<Player>()
        val db = readableDatabase
        var orderByQuery: String? = null
        when (orderValue) {
            ORDER_BY_ID -> orderByQuery = ORDER_BY + KEY_PLAYER_ID
            ORDER_BY_LEVEL -> orderByQuery = ORDER_BY + KEY_PLAYER_LEVEL
            ORDER_BY_STRENGTH -> orderByQuery = ORDER_BY + KEY_PLAYER_STRENGTH
            ORDER_BY_TOTAL -> orderByQuery = ORDER_BY + KEY_PLAYER_TOTAL
        }
        val playersTotalQuery = "SELECT " +
                KEY_PLAYER_ID + ", " +
                KEY_PLAYER_NAME + ", " +
                KEY_PLAYER_LEVEL + ", " +
                KEY_PLAYER_STRENGTH + ", " +
                KEY_PLAYER_COLOR + ", (" +
                KEY_PLAYER_LEVEL + " + " + KEY_PLAYER_STRENGTH + ") AS " + KEY_PLAYER_TOTAL +
                " FROM " + TABLE_PLAYERS + orderByQuery + " DESC"
        Log.v(TAG, playersTotalQuery)
        val cursor = db.rawQuery(playersTotalQuery, null)
        try {
            if (cursor!!.moveToFirst()) {
                do {
                    val player = Player(
                            id = cursor.getLong(cursor.getColumnIndex(KEY_PLAYER_ID)),
                            name = cursor.getString(cursor.getColumnIndex(KEY_PLAYER_NAME)),
                            color = cursor.getString(cursor.getColumnIndex(KEY_PLAYER_COLOR)),
                            levelScore = cursor.getInt(cursor.getColumnIndex(KEY_PLAYER_LEVEL)),
                            strengthScore = cursor.getInt(cursor.getColumnIndex(KEY_PLAYER_STRENGTH)),
                            totalScore = cursor.getInt(cursor.getColumnIndex(KEY_PLAYER_TOTAL))
                    )
                    playersList.add(player)
                } while (cursor.moveToNext())
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error while getting players")
        } finally {
            if (cursor != null && !cursor.isClosed) {
                cursor.close()
            }
        }
        return playersList
    }

    fun updatePlayer(player: Player): Player {
        val db = writableDatabase
        val values = ContentValues()
        values.put(KEY_PLAYER_LEVEL, player.levelScore)
        values.put(KEY_PLAYER_STRENGTH, player.strengthScore)
        db.update(TABLE_PLAYERS, values, KEY_PLAYER_ID + " = ?",
                arrayOf(player.id.toString()))
        return player
    }

    fun deletePlayer(id: Long) {
        Log.v(TAG, "Id of the player is " + id)
        val db = writableDatabase
        db.beginTransaction()
        try {
            db.delete(TABLE_PLAYERS, KEY_PLAYER_ID + " = ?", arrayOf(java.lang.Long.toString(id)))
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            Log.e(TAG, "Error while deleting player")
        } finally {
            db.endTransaction()
        }
    }

    private fun isTableEmpty(db: SQLiteDatabase, tableName: String): Boolean {
        val countQuery = "SELECT COUNT(*) FROM " + tableName
        val cursor = db.rawQuery(countQuery, null)
        cursor.moveToFirst()
        val count = cursor.getInt(0)
        if (count > 0) {
            return false
        } else {
            return true
        }
    }

    private fun addColorsToUpdatedPlayers(db: SQLiteDatabase) {
        val values = ContentValues()
        values.put(KEY_PLAYER_COLOR, "#3B1606")
        db.update(TABLE_PLAYERS, values, null, null)
    }

    fun clearPlayersStats() {
        val db = writableDatabase
        val values = ContentValues()
        values.put(KEY_PLAYER_LEVEL, 1)
        values.put(KEY_PLAYER_STRENGTH, 1)
        db.update(TABLE_PLAYERS, values, null, null)
    }

    fun insertStep(gameStep: GameStep) {
        val db = writableDatabase
        db.beginTransaction()
        try {
            val values = ContentValues()
            values.put(KEY_GAME_PLAYER_ID, gameStep.playerId)
            values.put(KEY_GAME_PLAYER_LEVEL, gameStep.playerLevel)
            values.put(KEY_GAME_PLAYER_STRENGTH, gameStep.playerStrength)
            db.insertOrThrow(TABLE_GAME, null, values)
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            Log.e(TAG, "Error while trying to insert step to database")
        } finally {
            db.endTransaction()
        }
    }

    fun clearSteps() {
        Log.v(TAG, "Clearing game steps")
        val db = writableDatabase
        db.beginTransaction()
        try {
            db.delete(TABLE_GAME, null, null)
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            Log.e(TAG, "Error while trying to delete game steps")
        } finally {
            db.endTransaction()
        }
    }

    val gameSteps: ArrayList<GameStep>
        get() {
            Log.v(TAG, "GET GAME STEPS")
            val gameSteps = ArrayList<GameStep>()
            val db = readableDatabase
            val lineChartQuery = "SELECT " +
                    KEY_GAME_PLAYER_ID + ", " +
                    KEY_GAME_PLAYER_LEVEL + ", " +
                    KEY_GAME_PLAYER_STRENGTH + ", " +
                    KEY_PLAYER_NAME + ", " +
                    KEY_PLAYER_COLOR + " FROM " + TABLE_GAME +
                    " INNER JOIN " + TABLE_PLAYERS + " ON players.id = " + KEY_GAME_PLAYER_ID
            Log.v("TAG", lineChartQuery)
            val cursor = db.rawQuery(lineChartQuery, null)
            try {
                if (cursor!!.moveToFirst()) {
                    do {
                        val gameStep = GameStep(
                        playerId = cursor.getLong(cursor.getColumnIndex(KEY_GAME_PLAYER_ID)),
                        playerLevel = cursor.getInt(cursor.getColumnIndex(KEY_GAME_PLAYER_LEVEL)),
                        playerStrength = cursor.getInt(cursor.getColumnIndex(KEY_GAME_PLAYER_STRENGTH)))
                        gameSteps.add(gameStep)
                    } while (cursor.moveToNext())
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error while getting game steps")
            } finally {
                if (cursor != null && !cursor.isClosed) {
                    cursor.close()
                }
            }
            return gameSteps
        }

    companion object {

        private val TAG = LogUtil.makeLogTag(MunchkinDatabaseHelper::class)

        private val DATABASE_NAME = "players_db"
        private val DATABASE_VERSION = 6

        private val TABLE_PLAYERS = "players"

        private val KEY_PLAYER_ID = "id"
        private val KEY_PLAYER_NAME = "name"
        private val KEY_PLAYER_LEVEL = "level"
        private val KEY_PLAYER_STRENGTH = "strength"
        private val KEY_PLAYER_COLOR = "color"
        private val KEY_PLAYER_TOTAL = "total"

        private val TABLE_GAME = "game"

        private val KEY_GAME_PLAYER_ID = "player_id"
        private val KEY_GAME_PLAYER_LEVEL = "player_level"
        private val KEY_GAME_PLAYER_STRENGTH = "player_strength"

        private val ORDER_BY = " ORDER BY "

        private val ORDER_BY_ID = 0
        private val ORDER_BY_LEVEL = 1
        private val ORDER_BY_STRENGTH = 2
        private val ORDER_BY_TOTAL = 3
    }

}
